### PR TITLE
chore: refactor actionability checks

### PR DESCRIPTION
**Pull request by dgozman**

### CODE REVIEW

File: src/server/dom.ts [hunk 1]
```diff
@@ -495,7 +499,7 @@ export class ElementHandle<T extends Node = Node> extends js.JSHandle<T> {
     return runAbortableTask(async progress => {
       progress.throwIfAborted();  // Avoid action that has side-effects.
       const poll = await this._evaluateHandleInUtility(([injected, node]) => {
-        return injected.waitForVisibleAndSelectText(node);
+        return injected.waitForElementStatesAndPerformAction(node, ['visible', 'editable'], injected.selectText.bind(injected));
```
Comment (aslushnikov) [inline src/server/dom.ts:502]: why do we wait for editable here?

File: src/server/dom.ts [hunk 2]
```diff
@@ -469,7 +473,7 @@ export class ElementHandle<T extends Node = Node> extends js.JSHandle<T> {
     return this._page._frameManager.waitForSignalsCreatedBy(progress, options.noWaitAfter, async () => {
       progress.log('  waiting for element to be visible, enabled and editable');
       const poll = await this._evaluateHandleInUtility(([injected, node, value]) => {
-        return injected.waitForEnabledAndFill(node, value);
+        return injected.waitForElementStatesAndPerformAction(node, ['visible', 'enabled', 'editable'], injected.fill.bind(injected, value));
```
Comment (aslushnikov) [inline src/server/dom.ts:476]: `'enabled'` is redundant given `'editable'`?

File: src/server/injected/injectedScript.ts [hunk 3]
```diff
@@ -297,148 +304,208 @@ export class InjectedScript {
...
-      select.value = undefined as any;
-      selectedOptions.forEach(option => option.selected = true);
-      progress.log('    selected specified option(s)');
-      select.dispatchEvent(new Event('input', { 'bubbles': true }));
-      select.dispatchEvent(new Event('change', { 'bubbles': true }));
-      return selectedOptions.map(option => option.value);
-    });
+  private _retarget(node: Node): Element | null {
+    let element = node.nodeType === Node.ELEMENT_NODE ? node as Element : node.parentElement;
+    if (element)
```
Comment (aslushnikov) [inline src/server/injected/injectedScript.ts:309]:
```js
if (!element)
  return null;
```

File: src/server/injected/injectedScript.ts [hunk 4]
```diff
@@ -297,148 +304,208 @@ export class InjectedScript {
...
+    if (state === 'hidden')
+      return !this.isVisible(element);
+
+    const disabled = ['BUTTON', 'INPUT', 'SELECT', 'TEXTAREA'].includes(element.nodeName) && element.hasAttribute('disabled');
+    if (state === 'disabled')
+      return disabled;
+    if (state === 'enabled')
+      return !disabled;
+
+    const editable = ['INPUT', 'TEXTAREA', 'SELECT'].includes(element.nodeName)
```
Comment (aslushnikov) [inline src/server/injected/injectedScript.ts:401]: editable should also work with `contenteditable`

File: src/server/injected/injectedScript.ts [hunk 5]
```diff
@@ -297,148 +304,208 @@ export class InjectedScript {
...
+      ? !element.hasAttribute('readonly')
+      : !!element.ownerDocument.defaultView!.getSelection();
```
Comment (aslushnikov) [inline src/server/injected/injectedScript.ts:403]: This looks sketchy: existence of selection doesn't define editability

File: src/server/injected/injectedScript.ts [hunk 6]
```diff
@@ -297,148 +304,208 @@ export class InjectedScript {
...
+    if (state === 'checked') {
+      if (element.getAttribute('role') === 'checkbox')
+        return element.getAttribute('aria-checked') === 'true';
+      if (element.nodeName !== 'INPUT')
+        return 'error:notcheckbox';
+      if (!['radio', 'checkbox'].includes((element as HTMLInputElement).type.toLowerCase()))
+        return 'error:notcheckbox';
+      return (element as HTMLInputElement).checked;
+    }
+    return false;
```
Comment (aslushnikov) [inline src/server/injected/injectedScript.ts:416]: nit: throw unknown state?

File: src/server/injected/injectedScript.ts [hunk 7]
```diff
@@ -456,69 +523,13 @@ export class InjectedScript {
     range.selectNodeContents(element);
     const selection = element.ownerDocument.defaultView!.getSelection();
     if (!selection)
-      return 'error:notvisible';
+      return 'error:notconnected';
```
Comment (aslushnikov) [inline src/server/injected/injectedScript.ts:526]: I guess this might also happen in case of selecting text in an element with `user-select: none;`

**Review by aslushnikov — APPROVED** [review]

File: src/server/dom.ts [hunk 8]
```diff
@@ -444,12 +446,14 @@ export class ElementHandle<T extends Node = Node> extends js.JSHandle<T> {
...
   async _selectOption(progress: Progress, elements: ElementHandle[], values: types.SelectOption[], options: types.NavigatingActionWaitOptions): Promise<string[] | 'error:notconnected'> {
-    const selectOptions = [...elements, ...values];
+    const optionsToSelect = [...elements, ...values];
     return this._page._frameManager.waitForSignalsCreatedBy(progress, options.noWaitAfter, async () => {
       progress.throwIfAborted();  // Avoid action that has side-effects.
       progress.log('  selecting specified option(s)');
       await progress.checkpoint('before');
-      const poll = await this._evaluateHandleInUtility(([injected, node, selectOptions]) => injected.waitForOptionsAndSelect(node, selectOptions), selectOptions);
+      const poll = await this._evaluateHandleInUtility(([injected, node, optionsToSelect]) => {
+        return injected.waitForElementStatesAndPerformAction(node, ['visible', 'enabled'], injected.selectOptions.bind(injected, optionsToSelect));
```
[the remark this answers is not in the document]
Comment (dgozman) [reply src/server/dom.ts:455]: I'll look into this in a follow up.

**Pull request merged**